corehq/warehouse/transforms/sql/user_location_dim.sql

```diff
@@ -0,0 +1,66 @@
+--- Delete all records that appear in the user staging table
+DELETE FROM {{ user_location_dim }} as ul_dim
+USING {{ user_staging }} as user_staging
+INNER_JOIN {{ user_dim }} as ud
+ON ud.user_id = user_staging.user_id
+
+WHERE ul_dim.user_dim = ud.id;
+
+
+--- Repopulate the UserLocationDim
+INSERT INTO {{ user_location_dim }} (
+       domain,
+       user_dim_id,
+       location_dim_id,
+       dim_last_modified,
+       dim_created_on,
+       deleted,
+       batch_id
+)
+--- webusers
+SELECT
+	domain,
+	user_dim_id,
+	location_dim_id,
+	now(),
+	now(),
+	false,
+	{{ batch_id }}
+FROM
+(
+	SELECT
+		user_dim.id as user_dim_id,
+		json_array_elements(json_array_elements(user_staging.domain_memberships::JSON) -> 'assigned_location_ids') as location_id,
+		json_array_elements(user_staging.domain_memberships::JSON) ->> 'domain' as domain
+		location_dim.id as location_dim_id
+	FROM {{ user_staging }} as user_staging
+	INNER JOIN {{ user_dim }} as user_dim
+	ON user_staging.user_id = user_dim.user_id
+	INNER JOIN {{ location_dim }} as location_dim
+	ON location_id = location_dim.location_id
+	WHERE user_staging.doc_type = 'WebUser' AND NOT user_staging.deleted
+)
+UNION
+--- mobile users
+SELECT
+	domain,
+	user_dim_id,
+	location_dim_id,
+	now(),
+	now(),
+	false
+	{{ batch_id }}
+FROM
+(
+	SELECT
+		user_dim.id as user_dim_id,
+		user_staging.domain as domain,
+		UNNEST(user_staging.assigned_location_ids) AS location_id
+		location_dim.id as location_dim_id
+FROM {{ user_staging }} AS user_staging
+INNER JOIN {{ user_dim }} AS user_dim
+ON user_staging.user_id = user_dim.user_id
+INNER JOIN {{ location_dim }} AS location_dim
```

Comment:
these inner joins would cause records to silently drop out instead of failing the process in case there's no match
 - i made a similar comment [here](https://github.com/dimagi/commcare-hq/pull/19054#commitcomment-27235402)

Reply:
That makes a lot of sense. I can work to fix that throughout these tables. Have you noticed it elsewhere besides here and the synclog

Reply:
i haven't seen it in any of the recent prs but it might be worth adding a task to the trello to just grep for all the joins we do and make sure they don't have this vulnerability.

adding a [balancing step](https://trello.com/c/tKhbykVj/22-relook-at-how-verified-works-for-committing-records) would also help catch these kinds of errors if they were to happen